verilog/tools/ls/autoexpand.cc

```diff
@@ -89,33 +154,156 @@ class AutoExpander {
+// ([^\s(]+?) – first group, at least one character other than whitespace or
+//              opening paren
+// \s*\(\s*   – optional whitespace, opening paren, optional whitespace again
+// ([^\s(]+?) – second group, same as the first one
+// \s*\)*     – optional whitespace, closing paren
+const std::regex AutoExpander::autotemplate_conn_re_{
+    R"(\.\s*([^\s(]+?)\s*\(\s*([^\s(]+?)\s*\))"};
+
+const AutoExpander::Template *AutoExpander::Template::from(
+    const Map &templates, const absl::string_view type_id,
+    const long int instance_offset) {
+  const auto templates_it = templates.find(type_id);
+  const Template *matching_tmpl = nullptr;
+  if (templates_it != templates.end()) {
+    // Linear search for the matching template (there should be very few
+    // templates per type, often just one)
+    for (const auto &tmpl : templates_it->second) {
+      if (instance_offset < tmpl.offset) break;
+      matching_tmpl = &tmpl;
+    }
+  }
+  return matching_tmpl;
+}
+
+void AutoExpander::Module::EmitPortDeclarations(
+    std::ostream &output, const absl::string_view indent,
+    const absl::string_view header,
+    const std::function<bool(const Port &)> &pred) const {
+  bool first = true;
+  for (const Port &port : ports_) {
+    if (!pred(port)) continue;
+    if (first) {
+      if (output.tellp() != 0) output << ',';
+      output << '\n' << indent << "// " << header << '\n' << indent;
+      first = false;
+    } else {
+      output << ", ";
+    }
+    output << port.name;
+  }
+}
+
+void AutoExpander::Module::EmitPortConnections(
+    std::ostream &output, const absl::string_view indent,
     const absl::string_view header,
-    const std::vector<absl::string_view> &port_names, std::ostream &output) {
-  if (port_names.empty()) return;
-  output << line_end_str_ << "// " << header << line_end_str_;
+    const std::function<bool(const Port &)> &pred, const Template *tmpl) const {
   bool first = true;
-  for (auto port : port_names) {
-    if (!first) {
-      output << ',' << line_end_str_;
+  for (const Port &port : ports_) {
+    if (!pred(port)) continue;
+    if (first) {
+      if (output.tellp() != 0) output << ',';
+      output << '\n' << indent << "// " << header;
+      first = false;
+    } else {
+      output << ',';
+    }
+    auto connected_name = port.name;
+    if (tmpl) {
+      auto it = tmpl->connections.find(port.name);
+      if (it != tmpl->connections.end()) connected_name = it->second;
     }
-    first = false;
-    output << port;
+    output << '\n'
+           << indent << '.' << port.name << '(' << connected_name << ")";
   }
 }
 
-absl::optional<AutoExpander::AutoComment> AutoExpander::FindInSpan(
-    const std::regex &re, const absl::string_view span) {
-  std::match_results<const char *> match;
+void AutoExpander::Module::RetrieveModuleHeaderPorts(const Symbol &module) {
+  const auto module_ports = GetModulePortDeclarationList(module);
+  if (module_ports) {
```

Comment:
Since this is embraching the whole rest of the function, reducing the cyclomatic complexity with an early out here will improve readability.

```c++
if (!module_ports) return;
```